README.md
[![Build Status](https://travis-ci.org/comtravo/ctparse.svg?branch=master)](https://travis-ci.org/comtravo/ctparse)

**This code is in early alpha stage. There can and will be potentially breaking changes right on the `master` branch**

# Comtravo Time Parser

## Background

The package `ctparse` is a pure python package to parse time expressions from natural language (i.e. strings). In many ways it builds on similar concepts as Facebook's `duckling` package (https://github.com/facebook/duckling). However, for the time being it only targets times and only German and English text.

In principle `ctparse` can be used to **detect** time expressions in a text, however its main use case is the semantic interpretation of such expressions. Detecting time expressions in the first place can - to our experience - be done more efficiently (and precisely) using e.g. CRFs or other models targeted at this specific task.

`ctparse` is designed with the use case in mind where interpretation of time expressions is done under the following assumptions:

* All expressions are relative to some pre-defined reference times
* Unless explicitly specified in the time expression, valid resolutions are in the future relative to the reference time (i.e. `12.5.` will be the next 12th of May, but `12.5.2012` should correctly resolve to the 12th of May 2012).
* If in doubt, resolutions in the near future are more likely than resolutions in the far future (not implemented yet, but any resolution more than i.e. 3 month in the future is extremely unlikely).

The specific comtravo use-case is resolving time expressions in booking requests which almost always refer to some point in time within the next 4-8 weeks.

`ctparse` currently is language agnostic and supports German and English expressions. This might get an extension in the future. The main reason is that in real world communication more often than not people write in one language (their business language) but use constructs to express times that are based on their mother tongue and/or what they believe to be the way to express dates in the target language. This leads to text in German with English time expressions and vice-versa. Using a language detection upfront on the complete original text is for obvious no solution - rather it would make the problem worse.

## Example

```python
from ctparse import ctparse
from datetime import datetime

# Set reference time
ts = datetime(2018, 3, 12, 14, 30)
ctparse('May 5th 2:30 in the afternoon', ts=ts)
```

This should return a `Time` object represented as `Time[0-29]{2018-05-05 14:30 (X/X)}`, indicating that characters `0-29` were used in the resolution, that the resolved date time is the 5th of May 2018 at 14:30 and that this resolution is neither based on a day of week (first `X`) nor a part of day (second `X`).

## Implementation
`ctparse` - as `duckling` - is a mixture of a rule and regular expression based system + some probabilistic modeling. In this sense it resembles a PCFG.

### Rules

At the core `ctparse` is a collection of production rules over sequences of regular expressions and (intermediate) productions.

Productions are either of type `Time` or type `Interval` and can have certain predicates (e.g. whether a `Time` is a part of day like `'afternoon'`).

A typical rule than looks like this:

```python
@rule(predicate('isDate'), dimension(Interval))
```

I.e. this rule is applicable when the intermediate production resulted in something that has a date, followed by something that is in interval (like e.g. in `'May 5th 9-10'`).

The actual production is a python function with the following signature:

```python
@rule(predicate('isDate'), dimension(Interval))
def ruleDateInterval(ts, d, i):
  """
  param ts: datetime - the current refenrence time
  d: Time - a time that contains at least a full date
  i: Interval - some Interval
  """
  if not (i.t_from.isTOD and i.t_to.isTOD):
    return None
  return Interval(
    t_from=Time(year=d.year, month=d.month, day=d.day,
                hour=i.t_from.hour, minute=i.t_from.minute),
    t_to=Time(year=d.year, month=d.month, day=d.day,
              hour=i.t_to.hour, minute=i.t_to.minute))
```

This production will return a new interval at the date of `predicate('isDate')` spanning the time coded in `dimension(Interval)`. If the latter does code for something else than a time of day (TOD), no production is returned, e.g. the rule matched but failed.

### Technical Background

Some observations on the problem:

- Each rule is a combination of regular expressions and productions.
- Consequently, each production must originate in a sequence of regular expressions that must have matched (parts of) the text.
- Hence, only subsequence of **all** regular expressions in **all** rules can lead to a successful production.

To this end the algorithm proceeds as follows:

1. Input a string and a reference time
2. Find all matches of all regular expressions from all rules in the input strings. Each regular expression is assigned an identifier.
3. Find all distinct sequences of these matches where two matches do not overlap nor have a gap inbetween
4. To each such subsequence apply all rules at all possible positions until no further rules can be applied - in which case one solution is produced

Obviously, not all sequences of matching expressions and not all sequences of rules applied on top lead to meaningful results. Here the **P**CFG kicks in:

- Based on example data (`corpus.py`) a model is calibrated to predict how likely a  production is to lead to a/the correct result. Instead of doing a breadth first search, the most promising productions are applied first.
- Resolutions are produced until there are no more resolutions or a timeout is hit.
- Based on the same model from all resolutions the highest scoring is returned.


HISTORY.rst
=======
History
=======

0.3.4 (2022-07-28)
------------------

* Add fuzzy matching on longer literals
* [internal] De-tangle corpus tests into isolated test cases
* Allow spaces around separators in ruleDDMMYYYY and ruleYYYYMMDD

0.3.3 (2022-07-18)
------------------

* Add rule for straight forward US formatted dates (`ruleYYYYMMDD`)
* Added rule `ruleYearMonth`
* Added corpus cases for some open issues that now pass
* Changed all internal imports to be absolute (i.e. `from ctparse.x` instead of `from .x`)
* Dropped `tox` (now using github actions)

0.3.2 (2022-07-18)
------------------

* Drop support for python 3.6, update dev requirements

0.3.1 (2021-07-07)
------------------

* Add support for python 3.9 on travis and in manifest; update build config

0.3.0 (2021-02-01)
------------------

* Removed latent rules regarding times (latent rules regarding dates are still present)
* Added latent_time option to customize the new behavior, defauld behavior is backwards-compatible

0.2.1 (2020-05-27)
------------------

* Update development dependencies
* Add flake8-bugbear and fixed issues

0.2.0 (2020-04-23)
------------------

* Implemented new type `Duration`, to handle lengths of time
* Adapted the dataset to include `Duration`
* Implemented basic rule to merge `Duration`, `Time` and `Interval` in simple cases.
* Created a make target to train the model `make train`

0.1.0 (2020-03-20)
------------------

* Major refactor of code underlying predictive model
* Based on a contribution from @bharathi-srini: replace naive bayes from sklearn by own implementation
* Thus remove dependencies on numpy, scipy, scikit-learn
* Predictions are much faster: 97/s in the old vs. 239/s in the new code base
* Performance identical
* Deprecate support for python 3.5, add 3.8
* Add more strict type checking rules (mypy.ini)
* Force black code formatting, make this a linter step, "black" all code

0.0.47 (2020-02-28)
-------------------

* Allow overlapping matches of regular expression when generating inital stack of "tokens"

0.0.46 (2020-02-26)
-------------------

* Implemented heuristics to detect (albeit imperfectly) military times

0.0.44 (2019-11-05)
-------------------

* Released time corpus
* Implemented training model using ctparse corpus

0.0.43 (2019-11-01)
-------------------

* Added slash as a general separator
* Added ruleTODTOD (to support expression like afternoon/evening)

0.0.42 (2019-10-30)
-------------------

* Removed nb module
* Fix for two digit years
* Freshly retrained model binary file

0.0.41 (2019-10-29)
-------------------

* Fix run_corpus refactoring bug
* Implemented retraining utilities

0.0.40 (2019-10-25)
-------------------

* update develop dependencies
* remove unused Protocol import from typing_extensions

0.0.39 (2019-10-24)
-------------------

* split ctparse file into several different modules
* added types to public interface
* introduced the Scorer abstraction to implement richer scoring strategies

0.0.38 (2018-11-05)
-------------------

* Added python 3.7 to supported versions (fix on travis available)

0.0.8 (2018-06-07)
------------------

* First release on PyPI.


ctparse/types.py
from datetime import datetime
from typing import Any, Dict, Optional, Tuple, Type, TypeVar

import regex
from regex import Regex
import enum

T = TypeVar("T", bound="Artifact")


class Artifact:
    def __init__(self) -> None:
        self.mstart = 0
        self.mend = 0
        self._attrs = ["mstart", "mend"]

    def update_span(self: T, *args: "Artifact") -> T:
        self.mstart = args[0].mstart
        self.mend = args[-1].mend
        return self

    def __len__(self) -> int:
        return self.mend - self.mstart

    def __bool__(self) -> bool:
        return True

    def __str__(self) -> str:
        return ""

    def __repr__(self) -> str:
        return "{}[{}-{}]{{{}}}".format(
            self.__class__.__name__, self.mstart, self.mend, str(self)
        )

    def nb_str(self) -> str:
        """Return a string representation without the bounds information."""
        return "{}[]{{{}}}".format(self.__class__.__name__, str(self))

    def __eq__(self, other: Any) -> bool:
        if type(other) != type(self):
            return False
        else:
            return all(getattr(self, a) == getattr(other, a) for a in self._attrs)

    def __hash__(self) -> int:
        return hash(tuple(getattr(self, a) for a in self._attrs))

    def _hasOnly(self, *args: str) -> bool:
        """check that all attributes set to True are set (i.e. not None) and
        all set to False are not set (i.e. None)

        """
        return all(
            getattr(self, a) is not None if a in args else getattr(self, a) is None
            for a in self._attrs
        )

    def _hasAtLeast(self, *args: str) -> bool:
        """check that all attributes set to True are set (i.e. not None) and
        all set to False are not set (i.e. None)

        """
        return all(getattr(self, a) is not None for a in args)


class RegexMatch(Artifact):
    def __init__(self, id: int, m: Regex) -> None:
        super().__init__()
        self._attrs = ["mstart", "mend", "id"]
        self.key = "R{}".format(id)
        self.id = id
        self.match = m
        self.mstart = m.span(self.key)[0]
        self.mend = m.span(self.key)[1]
        self._text = m.group(self.key)

    def __str__(self) -> str:
        return "{}:{}".format(self.id, self._text)


_pod_hours = {
    "earlymorning": {
        "offset": (4, 7),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "morning": {
        "offset": (6, 9),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "forenoon": {
        "offset": (9, 12),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "noon": {
        "offset": (11, 13),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "afternoon": {
        "offset": (12, 17),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "evening": {
        "offset": (17, 20),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "lateevening": {
        "offset": (18, 21),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "night": {
        "offset": (19, 22),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (-1, -1),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (1, 1),
        },
    },
    "first": {
        "offset": (0, 0),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (0, 0),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (0, 0),
        },
    },
    "last": {
        "offset": (23, 23),
        "early": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (0, 0),
        },
        "late": {
            "early": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "late": {"offset": (0, 0), "very": {"offset": (0, 0)}},
            "very": {"offset": (0, 0)},
            "offset": (0, 0),
        },
    },
}


def _mk_pod_hours() -> Dict[str, Tuple[int, int]]:
    def _add_ts(t1: Tuple[int, int], t2: Tuple[int, int]) -> Tuple[int, int]:
        return (t1[0] + t2[0], t1[1] + t2[1])

    def _mk(
        pod: str, pod_data: Dict[str, Any], t: Tuple[int, int]
    ) -> Dict[str, Tuple[int, int]]:
        r = {pod: _add_ts(t, pod_data["offset"])}
        for k, v in pod_data.items():
            if k == "offset":
                continue
            r.update(_mk(k + pod, v, r[pod]))
        return r

    res = {}
    for k, v in _pod_hours.items():
        if k == "offset":
            continue
        res.update(_mk(k, v, (0, 0)))
    return res


pod_hours = _mk_pod_hours()


_TIME_REGEX = regex.compile(
    r"(\d{4}|X)-(\d{2}|X)-(\d{2}|X) (\d{2}|X):(\d{2}|X) \((\d|X)\/(\w+)\)"
)


class Time(Artifact):
    def __init__(
        self,
        year: Optional[int] = None,
        month: Optional[int] = None,
        day: Optional[int] = None,
        hour: Optional[int] = None,
        minute: Optional[int] = None,
        DOW: Optional[int] = None,
        POD: Optional[str] = None,
    ) -> None:
        super().__init__()
        self._attrs = ["year", "month", "day", "hour", "minute", "DOW", "POD"]
        # Might add some validation here, did not to avoid the overhead
        self.year = year
        self.month = month
        self.day = day
        self.hour = hour
        self.minute = minute
        self.DOW = DOW
        self.POD = POD

    # -----------------------------------------------------------------------------
    # Make sure to not accidentially test bool(x) as False when x==0, but you meant
    # x==None
    # -----------------------------------------------------------------------------
    @property
    def isDOY(self) -> bool:
        """isDayOfYear <=> a dd.mm but not year"""
        return self._hasOnly("month", "day")

    @property
    def isDOM(self) -> bool:
        """isDayOfMonth <=> a dd but no month"""
        return self._hasOnly("day")

    @property
    def isDOW(self) -> bool:
        """isDayOfWeek <=> DOW is the 0=Monday index; fragile test, as the DOW
        could be accompanied by e.g. a full date etc.; in practice,
        however, the production rules do not do that.

        """
        return self._hasOnly("DOW")

    @property
    def isMonth(self) -> bool:
        return self._hasOnly("month")

    @property
    def isPOD(self) -> bool:
        """isPartOfDay <=> morning, etc.; fragile, tests only that there is a
        POD and neither a full date nor a full time
        """
        return self._hasOnly("POD")

    @property
    def isHour(self) -> bool:
        """only has an hour"""
        return self._hasOnly("hour")

    @property
    def isTOD(self) -> bool:
        """isTimeOfDay - only a time, not date"""
        return self._hasOnly("hour") or self._hasOnly("hour", "minute")

    @property
    def isDate(self) -> bool:
        """isDate - only a date, not time"""
        return self._hasOnly("year", "month", "day")

    @property
    def isDateTime(self) -> bool:
        """a date and a time"""
        return self._hasOnly("year", "month", "day", "hour") or self._hasOnly(
            "year", "month", "day", "hour", "minute"
        )

    @property
    def isYear(self) -> bool:
        """just a year"""
        return self._hasOnly("year")

    @property
    def hasDate(self) -> bool:
        """at least a date"""
        return self._hasAtLeast("year", "month", "day")

    @property
    def hasDOY(self) -> bool:
        """at least a day of year"""
        return self._hasAtLeast("month", "day")

    @property
    def hasDOW(self) -> bool:
        """at least a day of week"""
        return self._hasAtLeast("DOW")

    @property
    def hasTime(self) -> bool:
        """at least a time to the hour"""
        return self._hasAtLeast("hour")

    @property
    def hasPOD(self) -> bool:
        """at least a part of day"""
        return self._hasAtLeast("POD")

    def __str__(self) -> str:
        return "{}-{}-{} {}:{} ({}/{})".format(
            "{:04d}".format(self.year) if self.year is not None else "X",
            "{:02d}".format(self.month) if self.month is not None else "X",
            "{:02d}".format(self.day) if self.day is not None else "X",
            "{:02d}".format(self.hour) if self.hour is not None else "X",
            "{:02d}".format(self.minute) if self.minute is not None else "X",
            "{:d}".format(self.DOW) if self.DOW is not None else "X",
            "{}".format(self.POD) if self.POD is not None else "X",
        )

    @classmethod
    def from_str(cls: Type["Time"], text: str) -> "Time":
        match = _TIME_REGEX.match(text)
        if not match:
            raise ValueError("Invalid format")
        else:

            def parse_opt_int(x: str) -> Optional[int]:
                return None if x == "X" else int(x)

            pod = match.group(7)
            return cls(
                year=parse_opt_int(match.group(1)),
                month=parse_opt_int(match.group(2)),
                day=parse_opt_int(match.group(3)),
                hour=parse_opt_int(match.group(4)),
                minute=parse_opt_int(match.group(5)),
                DOW=parse_opt_int(match.group(6)),
                POD=None if pod == "X" else pod,
            )

    @property
    def start(self) -> "Time":
        if self.hour is None and self.hasPOD:
            hour = pod_hours[self.POD][0]  # type: ignore
        else:
            hour = self.hour or 0
        return Time(
            year=self.year,
            month=self.month,
            day=self.day,
            hour=hour,
            minute=self.minute or 0,
        )

    @property
    def end(self) -> "Time":
        if self.hour is None and self.hasPOD:
            hour = pod_hours[self.POD][1]  # type: ignore
        else:
            hour = self.hour if self.hour is not None else 23
        return Time(
            year=self.year,
            month=self.month,
            day=self.day,
            hour=hour,
            minute=self.minute if self.minute is not None else 59,
        )

    @property
    def dt(self) -> datetime:
        # Use the start time, in case we have a POD specification
        t = self.start
        if t.year is None or t.month is None or t.day is None:
            raise ValueError(
                "cannot convert under specified Time into datetime"
                ", missing at least one of year, month or day"
            )
        return datetime(t.year, t.month, t.day, t.hour or 0, t.minute or 0)


class Interval(Artifact):
    def __init__(
        self, t_from: Optional[Time] = None, t_to: Optional[Time] = None
    ) -> None:
        super().__init__()
        self._attrs = ["t_from", "t_to"]
        self.t_from = t_from
        self.t_to = t_to

    @property
    def isTimeInterval(self) -> bool:
        if self.t_from is None or self.t_to is None:
            return False
        else:
            return self.t_from.isTOD and self.t_to.isTOD

    @property
    def isDateInterval(self) -> bool:
        if self.t_from is None or self.t_to is None:
            return False
        return self.t_from.isDate and self.t_to.isDate

    def __str__(self) -> str:
        return "{} - {}".format(str(self.t_from), str(self.t_to))

    @classmethod
    def from_str(cls: Type["Interval"], text: str) -> "Interval":
        bounds = text.split(" - ")
        if len(bounds) != 2:
            raise ValueError("Invalid format")

        t_from = None if bounds[0] == "None" else Time.from_str(bounds[0])
        t_to = None if bounds[1] == "None" else Time.from_str(bounds[1])
        return cls(t_from=t_from, t_to=t_to)

    @property
    def start(self) -> Optional[Time]:
        if self.t_from is not None:
            return self.t_from.start
        else:
            return None

    @property
    def end(self) -> Optional[Time]:
        if self.t_to is not None:
            return self.t_to.end
        else:
            return None


@enum.unique
class DurationUnit(enum.Enum):
    MINUTES = "minutes"
    HOURS = "hours"
    DAYS = "days"
    NIGHTS = "nights"
    WEEKS = "weeks"
    MONTHS = "months"


class Duration(Artifact):
    def __init__(self, value: int, unit: DurationUnit):
        """Create a Duration using value and unit.

        Typical values for unit are:

        minute, hour, day, night, week, month, year
        """
        super().__init__()
        self.value = value
        self.unit = unit

    def __str__(self) -> str:
        return "{} {}".format(self.value, self.unit.value)

    @classmethod
    def from_str(cls: Type["Duration"], text: str) -> "Duration":
        value, unit = text.split()
        return Duration(int(value), DurationUnit(unit))
